File: iosApp/Well/SwiftUI/Screens/Call/CallScreen.swift
//
//  CallScreen.swift
//  Well
//
//  Created by Philip Dukhov on 12/29/20.
//  Copyright © 2020 Well. All rights reserved.
//

import SwiftUI
import SharedMobile

struct CallScreen: View {
    let state: CallFeature.State
    let listener: (CallFeature.Msg) -> Void

    var body: some View {
        GeometryReader { geometry in
            ZStack(alignment: .top) {
                let ongoing = state.status == .ongoing
                let bottomBarHeight = CallBottomBar.baseHeight + geometry.safeAreaInsets.bottom
                let callButtonsOffset = ongoing ? callButtonRadius - callButtonOffset : CallBottomBar.baseHeight
                GradientView(gradient: .callBackground)
                    .fillMaxSize()
                callInfoView(geometry)
                let minimizedBottomPadding = callButtonRadius * 2 + (ongoing ? bottomBarHeight - callButtonsOffset : geometry.safeAreaInsets.bottom)
                state.remoteVideoView.map {
                    videoView(
                        videoView: $0,
                        geometry: geometry,
                        minimizedBottomPadding: minimizedBottomPadding,
                        showFlipButton: false
                    )
                }
                state.localVideoView.map {
                    videoView(
                        videoView: $0,
                        geometry: geometry,
                        minimizedBottomPadding: minimizedBottomPadding,
                        showFlipButton: $0.position == .minimized
                    )
                }
                GeometryReader { geometry in
                    ZStack(alignment: .leading) {
                        DrawingContent(
                            state: state.drawingState,
                            enabled: state.controlSet == .drawing,
                            listener: {
                                listener(CallFeature.MsgDrawingMsg(msg: $0))
                            }
                        ).fillMaxSize()
                        .onAppear {
                            listener(
                                CallFeature.MsgDrawingMsg(
                                    msg: DrawingFeature.MsgUpdateLocalVideoContainerSize(size: geometry.size.toSize())
                                )
                            )
                        }
                    }.visible(state.drawingState.image == nil)
                }
                VStack(spacing: 0) {
                    Spacer()
                    if state.viewPoint == .mine {
                        HStack {
                            Spacer()
                            flipCameraButton()
                                .padding()
                                .offset(
                                    x: (controlMinSize -
                                        CallFeature
                                        .StateVideoViewPosition
                                        .minimized
                                        .sizeIn(geometry: geometry)
                                        .width) / 2,
                                    y: callButtonsOffset
                                )
                        }
                    }
                    callButtons()
                        .offset(y: callButtonsOffset)
                        .padding(.all, ongoing ? 0 : nil)
                    CallBottomBar(
                        state: state,
                        listener: listener,
                        geometry: geometry
                    ).frame(height: bottomBarHeight)
                    .offset(y: ongoing ? 0 : bottomBarHeight)
                } // VStack
                .frame(maxWidth: .infinity)
                .visible(state.controlSet == .call)
                
                DrawingPanel(
                    geometry: geometry,
                    state: state.drawingState,
                    listener: {
                        listener(CallFeature.MsgDrawingMsg(msg: $0))
                    }, back: {
                        listener(CallFeature.MsgBack())
                    }).visible(state.controlSet == .drawing)
            } // ZStack
                .edgesIgnoringSafeArea(.all)
        } // GeometryReader
            .foregroundColor(.white)
            .animation(.default)
    }

    @ViewBuilder
    private func videoView(
        videoView: CallFeature.StateVideoView,
        geometry: GeometryProxy,
        minimizedBottomPadding: CGFloat,
        showFlipButton: Bool
    ) -> some View {
        let fullscreen = videoView.position == .fullscreen
        HStack {
            if !fullscreen {
                Spacer() // move view to the right
            }
            VStack {
                if !fullscreen {
                    Spacer() // move view to the bottom
                }
                VideoView(context: videoView.context)
                    .frameInfinitable(size: videoView.position.sizeIn(geometry: geometry))
                if showFlipButton {
                    flipCameraButton()
                }
            }
        }.padding(.all, fullscreen ? 0 : nil)
        .padding(.bottom, fullscreen ? 0 : minimizedBottomPadding)
        .opacity(videoView.hidden ? 0 : 1)
    }

    @ViewBuilder
    private func callInfoView(
        _ geometry: GeometryProxy
    ) -> some View {
        let ongoing = state.status == .ongoing
        let profileImageSize = ongoing ? geometry.size : CGSize(size: geometry.size.width * 0.6)
        let profileImageOffset: CGFloat = ongoing ? 0 : (20 + geometry.safeAreaInsets.top)
        let textInfoOffset = ongoing ? 0 : profileImageOffset + profileImageSize.height

        ProfileImage(state.user, clipCircle: !ongoing)
            .clipped()
            .frame(size: profileImageSize)
            .offset(y: profileImageOffset)
            .edgesIgnoringSafeArea([])
        VStack {
            if ongoing {
                Spacer()
            }
            Text(state.user.fullName)
                .style(.h4)
                .multilineTextAlignment(.center)
                .frame(maxWidth: geometry.size.width * 0.9)
                .padding(.bottom, 11)
                .offset(y: ongoing ? -geometry.size.height / 4 : 0)
            if ongoing {
                Spacer()
            } else {
//            Label(
//                text: state.status.stringRepresentation,
//                style: .body1,
//                textColor: .white
//            )
                ViewContainer {
                    Text(state.status.stringRepresentation)
                        .frame(maxWidth: geometry.size.width * 0.9)
                }.frame(maxWidth: .infinity)
            }
        }.offset(y: textInfoOffset)
    }

    @ViewBuilder
    private func callButtons() -> some View {
        HStack {
            Spacer()
            CallDownButton {
                listener(CallFeature.MsgEnd())
            }
            switch state.status {
            case .incoming:
                Spacer()
                Spacer()
                CallUpButton {
                    listener(CallFeature.MsgAccept())
                }
            default:
                EmptyView()
            }
            Spacer()
        }
    }
    
    @ViewBuilder
    private func flipCameraButton() -> some View {
        Control(
            Image(systemName: "camera.rotate.fill")
                .font(.system(size: 30))
        ){
            listener(state.localDeviceState.toggleIsFrontCameraMsg())
        }
    }

    private let callButtonOffset = ViewConstants.CallScreenBottomBar().CallButtonOffset.toCGFloat()
    private let callButtonRadius = ViewConstants.CallScreen().CallButtonRadius.toCGFloat()
}

File: modules/androidUi/src/main/kotlin/com/well/modules/androidUi/composableScreens/call/_CallScreen.kt
package com.well.modules.androidUi.composableScreens.call

import com.well.modules.androidUi.components.GradientView
import com.well.modules.androidUi.components.ProfileImage
import com.well.modules.androidUi.components.controlMinSize
import com.well.modules.androidUi.composableScreens.call.drawing.DrawingContent
import com.well.modules.androidUi.composableScreens.call.drawing.DrawingPanel
import com.well.modules.androidUi.ext.thenIf
import com.well.modules.androidUi.ext.visibility
import com.well.modules.androidUi.ext.widthDp
import com.well.modules.features.call.callFeature.CallFeature.Msg
import com.well.modules.features.call.callFeature.CallFeature.State
import com.well.modules.features.call.callFeature.CallFeature.State.Status
import com.well.modules.models.User
import com.well.modules.utils.viewUtils.Gradient
import com.well.modules.utils.viewUtils.ViewConstants.CallScreen.BottomBar.CallButtonOffset
import com.well.modules.utils.viewUtils.ViewConstants.CallScreen.CallButtonRadius
import com.well.modules.features.call.callFeature.drawing.DrawingFeature.Msg as DrawingMsg
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.aspectRatio
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.navigationBarsPadding
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.statusBarsPadding
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material.MaterialTheme
import androidx.compose.material.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.RectangleShape
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.constraintlayout.compose.ConstrainedLayoutReference
import androidx.constraintlayout.compose.ConstraintLayout

@Composable
fun CallScreen(
    state: State,
    listener: (Msg) -> Unit,
) = ConstraintLayout(
    modifier = Modifier
        .fillMaxSize()
) {
    val ongoing = state.status == Status.Ongoing
    val callButtonOffset = if (ongoing) (CallButtonRadius - CallButtonOffset).dp else 0.dp

    //should be inside ConstraintLayoutScope and before it gets called
    fun Modifier.videoContainerModifier(
        position: State.VideoView.Position,
        reference: ConstrainedLayoutReference,
        bottomTopReference: ConstrainedLayoutReference,
    ) = constrainAs(reference) {
        when (position) {
            State.VideoView.Position.FullScreen -> Unit
            State.VideoView.Position.Minimized -> {
                bottom.linkTo(bottomTopReference.top)
                end.linkTo(parent.end)
            }
        }
    }
        .padding(
            when (position) {
                State.VideoView.Position.FullScreen -> {
                    0.dp
                }
                State.VideoView.Position.Minimized -> {
                    minimizedVideoContainerPadding
                }
            }
        )
        .offset(
            y = when (position) {
                State.VideoView.Position.FullScreen -> {
                    0.dp
                }
                State.VideoView.Position.Minimized -> {
                    callButtonOffset
                }
            }
        )

    val (localVideoView, remoteVideoView, profileImage, nameContainer, bottomView) = createRefs()

    GradientView(
        gradient = Gradient.CallBackground,
        modifier = Modifier
            .fillMaxSize()
    )
    ProfileImage(
        state.user,
        shape = if (ongoing) RectangleShape else CircleShape,
        aspectRatio = if (ongoing) null else 1f,
        modifier = Modifier
            .constrainAs(profileImage) {
                centerHorizontallyTo(parent)
            }
            .run {
                if (ongoing) {
                    fillMaxSize()
                } else {
                    fillMaxWidth(0.6f)
                        .padding(top = 30.dp)
                        .statusBarsPadding()
                }
            }
    )
    Column(
        modifier = Modifier
            .constrainAs(nameContainer) {
                centerHorizontallyTo(parent)
                if (ongoing) {
                    linkTo(parent.top, bottomView.top)
                } else {
                    top.linkTo(profileImage.bottom, margin = 20.dp)
                }
            }
            .fillMaxWidth(0.9f)
    ) {
        FullNameText(
            user = state.user,
            modifier = Modifier
                .fillMaxWidth()
        )
        if (!ongoing) {
            Text(
                state.status.stringRepresentation,
                color = androidx.compose.ui.graphics.Color.White,
                textAlign = TextAlign.Center,
                style = MaterialTheme.typography.body1,
                modifier = Modifier
                    .fillMaxWidth()
            )
        }
    }
    state.remoteVideoView?.apply {
        VideoViewContainer(
            view = this,
            onFlip = null,
            contentModifier = Modifier
                .videoModifier(position = position),
            modifier = Modifier.videoContainerModifier(
                position,
                reference = remoteVideoView,
                bottomTopReference = bottomView,
            )
        )
    }
    state.localVideoView?.apply {
        VideoViewContainer(
            view = this,
            onFlip = when (position) {
                State.VideoView.Position.FullScreen -> null
                State.VideoView.Position.Minimized -> {
                    { listener(state.localDeviceState.toggleIsFrontCameraMsg()) }
                }
            },
            contentModifier = Modifier
                .videoModifier(position = position),
            modifier = Modifier.videoContainerModifier(
                position,
                reference = localVideoView,
                bottomTopReference = bottomView,
            )
        )
    }
    DrawingContent(
        state.drawingState,
        listener = listener::invokeDrawingMsg,
        enabled = state.controlSet == State.ControlSet.Drawing,
        onSizeChanged = {
            listener.invokeDrawingMsg(DrawingMsg.UpdateLocalVideoContainerSize(it))
        },
        modifier = Modifier
            .fillMaxSize()
            .visibility(state.drawingState.image == null)
    )
    Column(
        horizontalAlignment = Alignment.CenterHorizontally,
        modifier = Modifier
            .constrainAs(bottomView) {
                bottom.linkTo(parent.bottom)
            }
            .fillMaxWidth()
            .thenIf(condition = !ongoing, Modifier.navigationBarsPadding())
            .visibility(state.controlSet == State.ControlSet.Call)
    ) {
        if (state.viewPoint == State.ViewPoint.Mine) {
            FlipCameraButton(
                onFlip = {
                    listener(state.localDeviceState.toggleIsFrontCameraMsg())
                },
                modifier = Modifier
                    .align(Alignment.End)
                    .padding(minimizedVideoContainerPadding)
                    .offset(
                        x = (controlMinSize - LocalContext.current.resources.displayMetrics.widthDp * minimizedVideoPart) / 2,
                        y = callButtonOffset
                    )
            )
        }
        Row(
            modifier = Modifier
                .padding(bottom = if (ongoing) 0.dp else 10.dp)
        ) {
            Spacer(modifier = Modifier.weight(1f))
            CallDownButton(
                modifier = Modifier
                    .offset(y = callButtonOffset)
            ) {
                listener.invoke(Msg.End)
            }
            when (state.status) {
                Status.Calling -> Unit
                Status.Incoming -> {
                    Spacer(modifier = Modifier.weight(2f))
                    CallUpButton {
                        listener.invoke(Msg.Accept)
                    }
                }
                Status.Connecting -> Unit
                Status.Ongoing -> Unit
            }
            Spacer(modifier = Modifier.weight(1f))
        }
        if (ongoing) {
            BottomBar(state, listener)
        }
    }
    DrawingPanel(
        state.drawingState,
        modifier = Modifier
            .visibility(state.controlSet == State.ControlSet.Drawing)
    ) {
        listener(Msg.DrawingMsg(it))
    }
}

@Composable
fun FullNameText(
    user: User,
    modifier: Modifier,
) = Text(
    user.fullName,
    color = androidx.compose.ui.graphics.Color.White,
    textAlign = TextAlign.Center,
    style = MaterialTheme.typography.h4,
    modifier = modifier
        .fillMaxWidth(0.9f)
)

private fun Modifier.videoModifier(
    position: State.VideoView.Position,
) = when (position) {
    State.VideoView.Position.FullScreen -> {
        fillMaxSize()
    }
    State.VideoView.Position.Minimized -> {
        fillMaxWidth(minimizedVideoPart)
            .aspectRatio(1080f / 1920)
    }
}

private const val minimizedVideoPart = 0.3f
private val minimizedVideoContainerPadding = 10.dp

private fun ((Msg) -> Unit).invokeDrawingMsg(msg: DrawingMsg) =
    invoke(Msg.DrawingMsg(msg))